overlord/snapstate/component.go

```diff
@@ -170,9 +170,18 @@ func doInstallComponent(st *state.State, snapst *SnapState, compSetup *Component
 		addTask(kmodSetup)
 	}
 
+	changingSnapRev := snapst.IsInstalled() && snapst.Current != snapSi.Revision
+
 	// We might be replacing a component
 	compInstalled := snapst.IsComponentInCurrentSeq(compSi.Component)
-	if compInstalled {
+
+	// note that we don't unlink the currect component if we're also changing
+	// snap revisions while installing this component. that is because we don't
+	// want to remove the component from the state of the previous snap revision
+	// (for the purpose of something like a revert). additionally, this is
+	// consistent with us keeping previous snap revisions mounted after changing
+	// their revision.
+	if compInstalled && !changingSnapRev {
```

Comment:
feels almost like it would be more readable to have the !changingSnapRev pred at the start